### PR TITLE
Remove 'CapacityBytes' from list of required parameters

### LINKED ISSUE #8955: CapacityBytes should not be a required parameters to create a volume with Redfish
### Summary

For the CreateVolume command of the `redfish_config` module, the documentation doesn't mention the fact that `CapacityBytes` is a required parameter of the command. 
But it is effectively required on the line #L3732 of redfish_utils.py.
https://github.com/ansible-collections/community.general/blob/9.4.0/plugins/module_utils/redfish_utils.py#L3732

So when i try to use this command without `CapacityBytes`, my task fail with :
```
"msg": "['RAIDType', 'Drives', 'CapacityBytes'] are required parameter to create a volume"
```

This parameter is not really needed and i can create a volume with curl without `CapacityBytes`

### Issue Type

Bug Report

### Component Name

redfish_config

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.15.3]
  ...
  python version = 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0] 
  jinja version = 3.1.2
```

### Community.general Version

```console (paste below)
$ ansible-galaxy collection list community.general
# /home/pyfontan/Dev/gitlab/ansible/ccin2p3/sysadmin_tools/.collections/ansible_collections
Collection        Version
----------------- -------
community.general 9.4.0
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```

### OS / Environment

Not relevant

### Steps to Reproduce

```yaml (paste below)
- name: Create Volume
  community.general.redfish_config:
    category: Systems
    command: CreateVolume
    baseuri: "{{ baseuri }}"
    username: "{{ username }}"
    password: "{{ password }}"
    storage_subsystem_id: "DE07A000"
    volume_details:
      Name: "SYSTEM"
      RAIDType: "RAID1"
      Drives:
        - "/redfish/v1/Systems/1/Storage/DE07A000/Drives/36"
        - "/redfish/v1/Systems/1/Storage/DE07A000/Drives/37"
```

### Expected Results

I expected the creation of the volume `SYSTEM` and the changed value of  :
```json
{                                                                                                          
    "changed": true,                                                                                                                                           
    "invocation": {                                                                                                                                            
        "module_args": {                                                       
            "auth_token": null,                                                
            "baseuri": "X.X.X.X",                                        
            "bios_attributes": {},                                             
            "boot_order": [],                                                  
            "category": "Systems",                                             
            "ciphers": null,                                                   
            "command": [                                                       
                "CreateVolume"                                                 
            ],                                                                 
            "hostinterface_config": {},                                        
            "hostinterface_id": null,                                          
            "network_protocols": {},                                           
            "nic_addr": "null",                                                
            "nic_config": {},                                                  
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",                                                                                                 
            "resource_id": null,                                               
            "secure_boot_enable": true,                                        
            "service_id": null,                                                
            "sessions_config": {},                                             
            "storage_subsystem_id": "DE07A000",                                                                                                                
            "strip_etag_quotes": false,                                        
            "timeout": 60,                                                     
            "username": "XXX",                                                                                                                       
            "volume_details": {                                                
                "Links": {                                                     
                    "Drives": [                                                
                        {                                                      
                            "@odata.id": "/redfish/v1/Systems/1/Storage/DE07A000/Drives/36"
                        },                                                     
                        {                                                      
                            "@odata.id": "/redfish/v1/Systems/1/Storage/DE07A000/Drives/37"
                        }                                                      
                    ]                                                          
                },                                                             
                "Name": "SYSTEM",                                              
                "RAIDType": "RAID1"                                            
            },                                                                 
            "volume_ids": []                                                   
        }                                                                      
    },                                                                         
    "msg": "Volume Created"                                                    
}    
```

### Actual Results

```json
{                                                                                                   
    "changed": false,                                                                                                                                          
    "invocation": {                                                                                                                                            
        "module_args": {                                                                                                                                       
            "auth_token": null,                                                                                                                                
            "baseuri": "X.X.X.X",                                                                                                                        
            "bios_attributes": {},                                                                                                                             
            "boot_order": [],                                                                                                                                  
            "category": "Systems",                                                                                                                             
            "ciphers": null,                                                                                                                                   
            "command": [                                                                                                                                       
                "CreateVolume"                                                                                                                                 
            ],                                                                                                                                                 
            "hostinterface_config": {},                                                                                                                        
            "hostinterface_id": null,                                                                                                                          
            "network_protocols": {},                                                                                                                           
            "nic_addr": "null",                                                                                                                                
            "nic_config": {},                                                                                                                                  
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",                                                                                                 
            "resource_id": null,                                                                                                                               
            "secure_boot_enable": true,                                                                                                                        
            "service_id": null,                                                                                                                                
            "sessions_config": {},                                                                                                                             
            "storage_subsystem_id": "DE07A000",                                                                                                                
            "strip_etag_quotes": false,                                                                                                                        
            "timeout": 60,                                                                                                                                     
            "username": "XXX",                                                                                                                       
            "volume_details": {                                                                                                                                
                "Drives": [                                                                                                                                    
                    "/redfish/v1/Systems/1/Storage/DE07A000/Drives/36",                                                                                        
                    "/redfish/v1/Systems/1/Storage/DE07A000/Drives/37"                                                                                         
                ],                                                                                                                                             
                "Name": "SYSTEM",                                                                                                                              
                "RAIDType": "Mirrored"                                                                                                                         
            },                                                                                                                                                 
            "volume_ids": []                                                                                                                                   
        }                                                                                                                                                      
    },                                                                                                                                                         
    "msg": "['RAIDType', 'Drives', 'CapacityBytes'] are required parameter to create a volume"                                                                 
}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by pyfontan**
##### SUMMARY

Make `CapacityBytes` an allowed parameter and not a required one.

Fixes #8955 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME

redfish_config

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by mraineri** [comment]
Why would a user not want to specify the size of the volume? When this was originally discussed, we did not think deferring to the service to dictate the capacity of the new volume is reliable and expect users to know how big of a volume to make in all cases.

**Comment by mraineri** [comment]
Technically "RAIDType" and "Drives" are not required per the schema either, but we had similar concerns with deferring to the service to determine the appropriate RAID type and which physical drives to use for the volume.

**Comment by pyfontan** [comment]
> Why would a user not want to specify the size of the volume? When this was originally discussed, we did not think deferring to the service to dictate the capacity of the new volume is reliable and expect users to know how big of a volume to make in all cases.
> Technically "RAIDType" and "Drives" are not required per the schema either, but we had similar concerns with deferring to the service to determine the appropriate RAID type and which physical drives to use for the volume.

@mraineri I think i do not totally understand your two comments.
User needs to specify on which drive he wants the raid to be created and which type of RAID. But the capacity can by default be the maximum size possible.
I don't think it's possible to create a volume with redfish with a payload not specifying RAIDType or Drive in it ?? Is it ?

**Comment by pyfontan** [comment]
Creation test with curl and a payload without RAIDType
```json
{
  "error": {
    "code": "Base.1.12.CreateFailedMissingReqProperties",
    "message": "See @Message.ExtendedInfo for more information.",
    "@Message.ExtendedInfo": [
      {
        "MessageId": "Base.1.12.CreateFailedMissingReqProperties",
        "Message": "The create operation failed because the required property %1 was missing from the request.",
        "Severity": "Warning",
        "Resolution": "Provide the missing required parameter.",
        "MessageArgs": [
          "RAIDType"
        ]
      }
    ]
  }
}
```

When asking for capabilities (extract):

```json
{
    "RAIDType@Redfish.RequiredOnCreate": true,
    "Links@Redfish.RequiredOnCreate": true,
    "Links": {
        "Drives@Redfish.RequiredOnCreate": true,
    }
}
```

**Comment by mraineri** [comment]
The issue is there's nothing in the Volume schema that says the service defaults to using the maximum available space. If we need that semantic, we need to provide that feedback to SNIA.

For Drives and RAIDType, today the Ansible module requires the user to provide this, but the schema definitions don't make these mandatory. The same issue applies in that the current schema definition does not dictate what the default behavior can be. The capabilities you've highlighted are provided by the vendor's service. If you look at the schema definitions (https://redfish.dmtf.org/schemas/swordfish/v1/Volume_v1.xml), there's nothing about which properties are required.

Technically, as far as the standard is concerned, this is a valid payload when creating a new Volume:

```
POST /redfish/v1/Systems/1/Storage/1/Volumes

{}
```

The resultant volume can have varying characteristics from different vendors. Vendor A can decide that it'll create a RAID0 volume that consumes all drives. Vendor B can decide to not use RAID at all and just allocate space from the first drive.

**Comment by mraineri** [comment]
> RAIDType and Link/Drives are, at least for my adapter, the only required parameters

Yes, that's a requirement imposed by the Redfish service you're using; it's not imposed by the standard.

**Comment by pyfontan** [comment]
> Yes, that's a requirement imposed by the Redfish service you're using; it's not imposed by the standard.

Ok, so CapacityBytes is a requirement imposed by the module that i don't understand.
And that requirement prevents the creation of volumes on HPE servers here and i think this is also the case for DELL servers.

**Comment by mraineri** [comment]
> Ok, so CapacityBytes is a requirement imposed by the module that i don't understand.

It's because the current definition in the Volume schema does not give any interoperable guidance on what the capacity should be if the user does not specify it. We decided to err on the side of caution for Ansible to ensure the user is acknowledging the size of the requested volume.

I can see it being reasonable that a service should create a volume that consumes all available space on the requested drives for the requested RAID type, but I would like others to weigh in on this assumption (perhaps even provide this as feedback to SNIA for adding to the definition of CapacityBytes for clarity).

**Comment by pyfontan** [comment]
> > Yes, that's a requirement imposed by the Redfish service you're using; it's not imposed by the standard.
> 
> Ok, so CapacityBytes is a requirement imposed by the module that i don't understand. And that requirement prevents the creation of volumes on HPE servers here and i think this is also the case for DELL servers.

Confirmed on DELL:
https://www.dell.com/support/manuals/fr-bi/idrac9-lifecycle-controller-v3.2-series/idrac_3.21.21.21_redfishapiguide/volume?guid=guid-30f08e85-68ad-43fa-9e7d-801f2a701780&lang=en-us

**Comment by pyfontan** [comment]
> The issue is there's nothing in the Volume schema that says the service defaults to using the maximum available space. If we need that semantic, we need to provide that feedback to SNIA.
> 
> For Drives and RAIDType, today the Ansible module requires the user to provide this, but the schema definitions don't make these mandatory. The same issue applies in that the current schema definition does not dictate what the default behavior can be. The capabilities you've highlighted are provided by the vendor's service. If you look at the schema definitions (https://redfish.dmtf.org/schemas/swordfish/v1/Volume_v1.xml), there's nothing about which properties are required.
> 
> Technically, as far as the standard is concerned, this is a valid payload when creating a new Volume:
> 
> ```
> POST /redfish/v1/Systems/1/Storage/1/Volumes
> 
> {}
> ```
> 
> The resultant volume can have varying characteristics from different vendors. Vendor A can decide that it'll create a RAID0 volume that consumes all drives. Vendor B can decide to not use RAID at all and just allocate space from the first drive.

I think that current schema definition don't make these mandatory because, it is made to create logical volumes over drive arrays. 
And what is mandatory for array creation is not for logical volume creation.

**Comment by pyfontan** [comment]
In the description of Redfish Volume 1.9.0 
https://www.dmtf.org/sites/default/files/standards/documents/DSP2046_2023.1.pdf

We can read at chapter 6.140.5.11 that RAIDType is a requested property.

**Comment by pyfontan** [comment]
Moreover, perhaps it's just a bug in the HPE redfish implementation but 'CapacityBytes' seems to be a readonly property :astonished: ...
And it's not really pertinent for this request , but if i try to create a volume with this payload :
```json
{
    "DisplayName": "SYSTEM",
    "RAIDType": "RAID1",
    "CapacityBytes": "107374182400",
    "Links": {
        "Drives": [
            {
                "@odata.id": "/redfish/v1/Systems/1/Storage/DE07A000/Drives/36"
            },
            {
                "@odata.id": "/redfish/v1/Systems/1/Storage/DE07A000/Drives/37"
            }
        ]
    }
}
```

The reply is :
```json
{
  "error": {
    "code": "iLO.0.10.ExtendedInfo",
    "message": "See @Message.ExtendedInfo for more information.",
    "@Message.ExtendedInfo": [
      {
        "MessageArgs": [
          "FIXME:NeedProperty"
        ],
        "MessageId": "iLO.2.23.PropertyNotWritableOrUnknown"
      }
    ]
  }
}
```

I'm a little bit surprised to be the first one to face this kind of problem.

**Comment by mraineri** [comment]
> In the description of Redfish Volume 1.9.0 https://www.dmtf.org/sites/default/files/standards/documents/DSP2046_2023.1.pdf
> 
> We can read at chapter 6.140.5.11 that RAIDType is a requested property.

That's just an example payload. It does not mean that "RAIDType" is required. In https://redfish.dmtf.org/schemas/swordfish/v1/Volume.v1_10_1.json, this is the list of required properties for the Volume resource:

```
            "required": [
                "@odata.id",
                "@odata.type",
                "Id",
                "Name"
            ],
```

The schema for Volume also does not include the `requiredOnCreate` term, meaning the client is not required to supply any properties.

**Comment by mraineri** [comment]
> Moreover, perhaps it's just a bug in the HPE redfish implementation but 'CapacityBytes' seems to be a readonly property 😲

Redfish services are not required to support write functions for all writable properties; if they are constrained in some manner (or have no requirements to support the capability), they are free to make writable properties read-only per the Redfish spec:

> A service may implement a writable property as read-only.

**Comment by mraineri** [comment]
@pyfontan after talking about this with a few others, we did agree that it should be okay from an Ansible perspective to remove CapacityBytes as a required input parameter. We can see that it might be difficult in many cases for a user to determine the appropriate capacity size of the drive.

One thing we'd like to see is at least some description text for the parameter to specify that if the capacity is not specified, the Redfish service will be the one to determine the size to allocate, and depending on the design of the service, this might not always be the maximum available size. Would you be able to add this description text?

File: changelogs/fragments/8956-remove-capacitybytes-from-the-required-parameters_list.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - redfish_confg - remove ``CapacityBytes`` from required paramaters of the CreateVolume command (https://github.com/ansible-collections/community.general/pull/8956).
```
Comment (russoz) [inline changelogs/fragments/8956-remove-capacitybytes-from-the-required-parameters_list.yml:2]:
```suggestion
  - redfish_confg - remove ``CapacityBytes`` from required paramaters of the ``CreateVolume`` command (https://github.com/ansible-collections/community.general/pull/8956).
```

File: plugins/modules/redfish_config.py [hunk 2]
```diff
@@ -164,6 +164,9 @@
     required: false
     description:
       - Setting dict of volume to be created.
+      - If CapacityBytes key is not specified in this dictionary, the size of
+        the volume will be determined by the Redfish service. It's possible the
```
Comment (felixfontein) [inline plugins/modules/redfish_config.py:168]:
```suggestion
      - If C(CapacityBytes) key is not specified in this dictionary, the size of
        the volume will be determined by the Redfish service. It is possible the
```

**Comment by felixfontein** [comment]
@pyfontan thanks for your contribution!
@mraineri @russoz thanks for reviewing!

**Pull request merged**